### ISSUE TITLE
Figure 2 플롯 방법

### ISSUE THREAD

**Issue by shlee625** [baiksung/L2F#2]
안녕하세요. 먼저 코드를 공유해주셔서 감사합니다.
논문을 보면 [26] 논문에서 보여준 것처럼 Figure 2에서 loss landscape과 gradient predictiveness, 그리고 ꞵ-smoothness에 대한 plot을 확인할 수가 있는데요. 해당 plot들을 어떤 방식으로 구현을 하셨는지 여쭙고 싶습니다. 

1. 각각의 inner-loop step에서 구한 gradient 방향에 대해 learning rate를 조정하여 여러 weight를 구한 다음 각각에 대한 loss 및 gradient를 구해서 plot하는게 맞나요?
2. Appendix A에서 'We take an average of these values over the number of inner-loop updates and plot them against training iterations'로 표현을 하셨는데, 이는 1에서 조정된 여러 learning rate에 대해 inner-loop step (코드에서는 5회)에 대해 averaging한 것으로 이해하면 될까요?
3. 1번의 방식이 맞다면, learning rate 조정의 범위는 어떻게 설정하셨는지 여쭙고 싶습니다. [26] 논문을 보면 Appendix A.2.2에서 [1/2, 4] X step size로 스텝 길이를 조정했다고 나와 있는데, 해당 범위 내에서도 선택할 수 있는 값들이 많기 때문에 어떠한 방식으로 진행하셨는지 궁금합니다.

건강 조심하세요. 감사합니다.

**Comment by baiksung** [comment]
안녕하세요,

남기신 질문에 답변 드립니다.

1. 네 맞습니다. [26] 논문에서처럼 구한 gradient 방향에서 step size를 조정하면서 loss과 gradient variation 을 구합니다. [26] 논문의 Section 3.2도 참고하시면 좋을 것 같습니다.
2. 네 inner-loop step에 대해 averaging 하였습니다.
3. 저희도 [26] 논문에서는 선택한 값을 찾을 수 없어서 [1/2, 4] 에서 0.5간격 (0.5, 1.0, ...) 으로 스텝 길이를 조정했습니다.

[26] Santurkar et al. How Does Batch Normalization Help Optimization? 

궁금하신 점이 해결됐길 바랍니다.
건강 유의하시길 바랍니다.

감사합니다.

**Comment by shlee625** [comment]
@baiksung 답변해주셔서 감사합니다. 많은 도움이 되었습니다.

**Comment by shlee625** [comment]
안녕하세요.
plot 방법과 관련하여 몇 가지 더 여쭤보고 싶어 글을 남깁니다.

1. gradient predictiveness를 계산할 때 두 gradient간의 L2-distance를 계산하는데요. 하나는 기존의 training 방식으로 구한 gradient G1입니다. 이후 G1 방향으로 여러 learning rate로 update한 weight W2를 구하는데, 다른 하나는 W2를 통해 계산한 loss L2의 W2에 대한 gradient인가요, 아니면 W1에 대한 gradient인가요? 

2. beta-smoothness의 경우 위에서 구한 gradient difference의 maximum 값을 이용하셨다고 하셨는데 over the distance라는 부분이 조금 헷갈립니다. 여기서 말하는 distance는 기존의 weight W1과 특정 learning rate로 update된 weight W2 간의 distance, 즉 learning rate * G1의 L2-norm일텐데 이를 maximum gradient difference에 나눠준 값을 beta로 두고 plot하신 건가요?

3. batch마다 2개의 task가 학습때 이용이 되는데 task마다 따로 loss, gradient difference, beta를 구하고 task-wise로 averaging하고 다시 inner-loop step별로 하셨는지, 아니면 task 구분없이 step별로 구한 다음 평균내신건지 여쭙고 싶습니다.


질문이 많았습니다. 바쁘신 가운데 감사드립니다.

**Comment by baiksung** [comment]
안녕하세요,

논문 준비로 경황이 없어서 이제야 답변을 드립니다.

1. [26] Section 3.2에서 설명한대로 gradient predictiveness을 계산할 때 G1방향으로 여러 learning rate로 update한 각 W2 위치에서의 gradient와 기존 위치에서의 gradient G1를 비교하여서 계산하였습니다.

2. 이 부분도 [26]와 말씀하신 것과 마찬가지로 maximum gradient difference을 learning rate * G1의 L2-norm로 나눈 값으로 계산하였습니다.

3. task-wise로 averaging하여서 plot 하였습니다.

궁금하신 점이 해결됐길 바랍니다.

**Comment by shlee625** [comment]
감사합니다.
